src/RoyalLibrary/SortingExtensions.cs

```diff
@@ -0,0 +1,166 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+namespace ByteDecoder.RoyalLibrary
+{
+
+    /// <summary>
+    /// Advanced sorting for collections
+    /// </summary>
+    public static class SortingExtensions
+    {
+
+
+        /// <summary>
+        /// returns a new sorted ICollection 
+        /// </summary>
+        /// <param name="source"></param>
+        /// <returns>An Enumerable sorted collection by LastName</returns>
+        public async static Task<ICollection<string>> RoyalSortLastAsync(this IEnumerable<string> source)
+        {
+            int rowsCount = source.Count();
+            byte lastspaceIndex = 0;
+            string firstName, lastName = null;
+            var tempDic = new MultiKeysValue() { Capacity = rowsCount };
+            var tempList = new List<string>(rowsCount);
+            var tempResult = new List<string>(rowsCount);
+
+            await Task.Run(() =>
+            {
+                foreach (var item in source)
+                {
+
+                    var spaces = item.Count(char.IsWhiteSpace);
+                    lastspaceIndex = (byte)item.IndexOfNth(' ', spaces - 1);
+
+                    //get the whole firstname(s) , then the lastName, add them to tempDic which will be used
+                    //to get firstname(s) by lastName
+                    //lastname will be added to the temp list and sorted everytime 
+                    firstName = item.Substring(0, lastspaceIndex);
+                    var lastNameLength = item.Length - firstName.Length;
+                    lastName = item.Substring(lastspaceIndex, lastNameLength);
+
+                    tempDic.Add(firstName, lastName);
+                    tempList.Add(lastName);
+                    tempList.Sort();
+
+
+                }
+            }).ContinueWith(async (filtering) =>
+            {
+                foreach (var LastName in tempList)
+                {
+                    //get the firstName(s) from tempDic by lastName
+
+                    var dicfirstName = tempDic.AsParallel().First(x => x.Value == LastName).Key;
+                    await Task.Delay(0);//fake await to run everything smoothly
+                                        //combine names again with sorted
+                    tempResult.Add(dicfirstName + LastName);
+
+                }
+            });
+            tempDic = null;
+            tempList = null;
+            return tempResult;
+
+
+
+
+
+        }
+        /// <summary>
+        /// returns a new sorted ICollection 
+        /// </summary>
+        /// <param name="source"></param>
+        /// <param name="action"></param>
+        /// <returns>An Enumerable sorted collection by LastName</returns>
+        public async static Task<ICollection<string>> RoyalSortLastAsync(this IEnumerable<string> source, Action<string> action)

```

Comment:
Another thing that we can do here, is taking the *RoyalSortLastAsync* method overloading advantage to have this second method implementation in this way unless if you think the behavior  will affect the output:

`        
public async static Task<ICollection<string>> RoyalSortLastAsync(this IEnumerable<string> source, Action<string> action)
        {
            var tempResult = await source.RoyalSortLastAsync();
            
            foreach (var item in tempResult.AsParallel())
            {
                action.Invoke(item);
            }

            return tempResult;
        }
`

Reply:
The idea is nice and I thought about it at first, but **unfortunately** this will not gonna work to achieve the **right** output, I mean the method is assumed working well with **strings**, so defining the generic type must be of type string to avoid errors. If i had assumed working with non strings such as classes or numbers it would have failed.

List<int> can be easy sorted or reversed with `MyList.Sort()` or` MyList.Reverse() ` , 
But the big problem  working with dynamic classes, I mean.... List<User> , List<Player> , each of them will have unique properties, looping over them `foreach(item in ....)` will return the namespace.className.

and the type of action I had used in parameter , it will help to treat each item individual .

```
list.RoyalSortLastAsync(item=>
{
  //code with each single item
});
```
is alternative to
`foreach(item in RoyalSortLastAsync) //some code`

but the action is supposed to work AsParallel to achieve the best performance,

Reply:
Sounds good